### ISSUE TITLE
Comments not emitted in chaining calls

### ISSUE THREAD

**Issue by tnhu** [jlongster/prettier#200]
before:

```js
request
  .post('/version')  // comment 1
  .set('Prefer', 'plurality=singular') /* comment 2 */
  .send()
  .end((error, response) => {});
```

after:

```js
request
  .post('/version')
  .set('Prefer', 'plurality=singular')
  .send()
  .end((error, response) => {});
```

Related issues:

https://github.com/jlongster/prettier/issues/31
https://github.com/jlongster/prettier/issues/87

**Comment by vjeux** [comment]
This most likely has to do with `printMemberChain` not using the generic way of printing things. Comments are preserved if there are no function calls.

```js
request
  .post('/version')  // comment 1
  .set('Prefer', 'plurality=singular') /* comment 2 */
  .send()
  
request
  .post.version  // comment 1
  .set.Prefer.plurality.singular /* comment 2 */
  .send()
```

```js
request.post("/version").set("Prefer", "plurality=singular").send();

// comment 1
request.post.version.set.Prefer.plurality.singular /* comment 2 */.send();
```

https://jlongster.github.io/prettier/#%7B%22content%22%3A%22request%5Cn%20%20.post('%2Fversion')%20%20%2F%2F%20comment%201%5Cn%20%20.set('Prefer'%2C%20'plurality%3Dsingular')%20%2F*%20comment%202%20*%2F%5Cn%20%20.send()%5Cn%20%20%5Cnrequest%5Cn%20%20.post.version%20%20%2F%2F%20comment%201%5Cn%20%20.set.Prefer.plurality.singular%20%2F*%20comment%202%20*%2F%5Cn%20%20.send()%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Afalse%7D%7D

**Comment by jlongster** [comment]
Moved to #275 to consolidate all the comment bugs

**Comment by jlongster** [comment]
Re-opening as I didn't fix this one in my comment overhaul.

**Comment by dmnd** [comment]
[Related](https://jlongster.github.io/prettier/#%7B%22content%22%3A%22request%5Cn%20%20%2F%2F%20handle%20escaped%20backticks%5Cn%20%20.post('%2Fversion')%5Cn%20%20%2F*%20comment%202%20*%2F%5Cn%20%20.set('Prefer'%2C%20'plurality%3Dsingular')%5Cn%20%20.send()%5Cn%20%20%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Afalse%2C%22doc%22%3Afalse%7D%7D). Comments on their own lines really mess with the output.

```js
request
  // handle escaped backticks
  .post('/version')
  /* comment 2 */
  .set('Prefer', 'plurality=singular')
  .send()
```

```js
request.// handle escaped backticks
post("/version")./* comment 2 */
set("Prefer", "plurality=singular").send();
```

I'm not sure if this belongs in here or in its own issue.

**Comment by vjeux** [comment]
Thanks @dmnd, this is a known issue and there's a fix in https://github.com/jlongster/prettier/pull/469

Also, awesome job on dedent package, it makes template literals usable!

**Comment by amasad** [comment]
Is this the same issue:

```js
str.replace(/\//g, '_')
  // Remove ending '=' 15 bits doesn't have it but just to be safe
  .replace(/=+$/, '');
```

prints:
```js
str
  .replace(/\//g, "_")
  .// Remove ending '=' 15 bits doesn't have it but just to be safe
  replace(/=+$/, "");
```

**Comment by amasad** [comment]
Actually this reproduces on the prettier codebase:

```
$ npm run format:all
$ git diff src/printer.js

 function printNumber(rawNumber) {
-  return rawNumber.toLowerCase()
-    // Remove unnecessary plus and zeroes from scientific notation.
-    .replace(/^([\d.]+e)(?:\+|(-))?0*/, "$1$2")
-    // Make sure numbers always start with a digit.
-    .replace(/^\./, "0.")
-    // Remove trailing dot.
-    .replace(/\.(?=e|$)/, "");
+  return rawNumber
+    .toLowerCase()
+    .// Remove unnecessary plus and zeroes from scientific notation.
+    replace(/^([\d.]+e)(?:\+|(-))?0*/, "$1$2")
+    .// Make sure numbers always start with a digit.
+    replace(/^\./, "0.")
+    .// Remove trailing dot.
+    replace(/\.(?=e|$)/, "");
 }
```

**Comment by vjeux** [comment]
@amasad for this, I have a fix that I need to properly extract out before we can release it. I want it in before I do the next release

```diff
diff --git a/src/comments.js b/src/comments.js
index e05c7f5..d31122e 100644
--- a/src/comments.js
+++ b/src/comments.js
@@ -152,7 +152,10 @@ function attach(comments, ast, text) {
       // If a comment exists on its own line, prefer a leading comment.
       // We also need to check if it's the first line of the file.

-      if (followingNode) {
+      if (enclosingNode.type === "MemberExpression" &&
+          followingNode.type === "Identifier") {
+        addLeadingComment(enclosingNode, comment);
+      } else if (followingNode) {
         // Always a leading comment.
         addLeadingComment(followingNode, comment);
       } else if (precedingNode) {
```
